packages/aws_common/lib/src/http/aws_http_client_io.dart

```diff
@@ -0,0 +1,500 @@
+              Header(utf8.encode(entry.key), utf8.encode(entry.value)),
+        ],
+      );
+
+      var requestBytesRead = 0;
+      request.body
+          .map((chunk) {
+            requestBytesRead += chunk.length;
+            requestProgress.add(requestBytesRead);
+            return DataStreamMessage(chunk);
+          })
+          .takeUntil(cancelTrigger)
+          .listen(
+            stream.outgoingMessages.add,
+            onDone: () {
+              requestProgress.close();
+              stream.outgoingMessages.close();
+            },
+          );
+
+      final gotHeaders = Completer<void>();
+      onCancel = () {
+        logger.verbose('onCancel triggered');
+        if (!gotHeaders.isCompleted) {
+          gotHeaders.completeError(const CancellationException());
+        }
+      };
+      stream.onTerminated = (code) {
+        logger.debug('Stream terminated: $code');
+        _http2Connections.remove(uri.authority);
+      };
+
+      final headers = CaseInsensitiveMap<String>({});
+      final bodyController = StreamController<List<int>>(sync: true);
+
+      late final StreamSubscription<StreamMessage> subscription;
+      subscription = stream.incomingMessages.listen(
+        (message) {
+          if (completer.isCanceled) {
+            subscription.cancel();
+            return;
+          }
+          if (message is HeadersStreamMessage) {
+            headers.addEntries(
+              message.headers.map((header) {
+                return MapEntry(
+                  utf8.decode(header.name),
+                  utf8.decode(header.value),
+                );
+              }),
+            );
+            if (message.endStream) {
+              gotHeaders.complete();
+            }
+          } else if (message is DataStreamMessage) {
+            if (!gotHeaders.isCompleted) {
+              gotHeaders.complete();
+            }
+            bodyController.add(message.bytes);
+          }
+        },
+        onError: (Object error, StackTrace stackTrace) {
+          logger.debug('Error in stream: $error');
+          if (!gotHeaders.isCompleted) {
+            gotHeaders.completeError(
+              AWSHttpException(request, error),
+              stackTrace,
+            );
+            return;
+          }
+          if (!bodyController.isClosed) {
+            bodyController
+              ..addError(AWSHttpException(request, error), stackTrace)
+              ..close();
+          }
+        },
+        onDone: () {
+          logger.verbose('Stream done');
+          if (!gotHeaders.isCompleted) {
+            gotHeaders.completeError('No headers received');
+            return;
+          }
+          bodyController.close();
+        },
+      );
+      logger.verbose('Subscription established');
+      onCancel = () {
+        logger.verbose('onCancel triggered');
+        subscription.cancel();
+        transport.terminate();
```

Comment:
question: Would this interrupt/close any other in-flight requests on this client which resolve to the same authority?

Reply:
Yep good catch! Updated and added tests